keps/sig-node/3815-pod-condition-for-configuration-errors/README.md

```diff
@@ -0,0 +1,1021 @@
+-->
+
+As a batch user, it is very likely that users experience configuration errors when submitting pods to Kubernetes.  We would like to propose a new condition so that users can determine if a pod has an error in startup.  A new condition will provide a quick way to verify if a pod detected a unrecoverable configuration error.
+
+## Motivation
+
+Batch users may not have deep knowledge of Kubernetes.  It is common occurrence to find users submitting Pods with configuration errors (invalid images, wrong image names, missing volumes, missing secrets, wrong volume/secret for config map).  These cases cause the Pod to be stuck in pending and there is code required to remove these pods or user input to remove.
+
+As a batch system developer, we find that higher level controllers have to write code to handle Pending Pod cases.  Since the pods are not considered failed, the higher level controller needs to transition this state to failed in order to enforce retries.  For example, the job controller will run these pods and the state of the Pod will be in Pending.  The job controller does not transition this state to failed even if the pod will never start running.
+
+For CNCF projects, in the Armada project, we implement logic that determines retry ability based on events and statuses.  Events are not standard so we expect issues when updating and we discovered various states that don't have a representative status.  
+
+It would be ideal to have a standard condition for these common configuration errors.
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+- Add a new condition PodFailedToStartup to detect failures in configuration.
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+- Will not modify behavior for ImagePullBackOff
+   - This status can be business as usual or image pull errors
+   - Unclear if it is possible to detect this as a failure
+   - Invalid ImagePullSecrets are a common fail case but it is difficult to 
+- Configuration errors that happen during ContainerCreation (Mounting volumes) will not be included as part of this condition
+   - PodReadyToStart condition actually covers these cases so there is no need to cover them in this KEP.
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+- Add a PodFailedToStartup Condition for the following cases:
+
+  - Invalid Image Name
+  - Image not found locally when ImagePullPolicy is set to never.
```

Comment:
This is represented in `containerStatuses` but a `condition` seems appropriate

Reply:
Yea I think all these cases are represented in `containerStatuses` but I think moving to a condition would help the community.  Mostly because the only way to determine this when put to loop through the containers and check for particular statuses.  

A big part of this KEP is to standarize on at least a single condition so controllers and third party developers can "fix" or automate fixing of common problems related to pods.